clippy_lints/src/unnecessary_vec_drain.rs

```diff
@@ -0,0 +1,96 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::source::snippet_with_applicability;
+use clippy_utils::ty::{is_type_diagnostic_item, is_type_lang_item};
+use rustc_ast::LitKind::Int;
+use rustc_errors::Applicability;
+use rustc_hir as hir;
+use rustc_hir::{LangItem, Stmt};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::symbol::sym;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for usage of `vec.drain(..)` with RangeFull that is not bound to a value `let`.
+    ///
+    /// ### Why is this bad?
+    /// This creates an iterator that is dropped immediately.
+    ///
+    /// `vec.clear()` also shows clearer intention.
+    ///
+    /// ### Example
+    /// ```rust
+    /// let mut vec: Vec<i32> = Vec::new();
+    ///  vec.drain(..);
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// let mut vec: Vec<i32> = Vec::new();
+    /// vec.clear();
+    /// ```
+    #[clippy::version = "1.66.0"]
+    pub UNNECESSARY_VEC_DRAIN,
+    style,
+    "unnecessary `vec.drain(..)`"
+}
+declare_lint_pass!(UnnecessaryVecDrain => [UNNECESSARY_VEC_DRAIN]);
+
+impl LateLintPass<'_> for UnnecessaryVecDrain {
+    fn check_stmt<'tcx>(&mut self, cx: &LateContext<'tcx>, stmt: &'tcx Stmt<'tcx>) {
+        if let hir::StmtKind::Semi(semi_expr) = &stmt.kind {
+            if let hir::ExprKind::MethodCall(path, rcvr, method_args,drain_span) = &semi_expr.kind
+                && path.ident.name == sym!(drain)
+            {
+                let ty = cx.typeck_results().expr_ty(rcvr);
+                if let [expr_element] = &**method_args
+                    && is_type_diagnostic_item(cx, ty, sym::Vec)
```

Comment:
You're combining `if let` chains with the `if_chain` macro. Please use `if let` chains everywhere. 

You can also use irrefutable patterns in `if let` chains, so this can be all in one chain up until here.